### PR TITLE
add experimental SQL expressions

### LINKED ISSUE #2215: basic SQL expressions
Z should be able to embed SQL expressions.  We will start with a basic implementation of SELECT, FROM, JOIN, GROUP BY, HAVING, ORDER BY, and LIMIT.  Support for recursive joins will be added later.

### PR DESCRIPTION
**Pull request by mccanne**
Closes #2215

### CODE REVIEW

File: field/field.go [hunk 1]
```diff
@@ -56,6 +56,28 @@ func (f Static) HasPrefix(prefix Static) bool {
 	return len(f) >= len(prefix) && prefix.Equal(f[:len(prefix)])
 }
 
+func (f Static) IsPrefixOf(of Static) bool {
+	return of.HasPrefix(f)
+}
+
```
Comment (nwt) [inline field/field.go:62]: `HasPrefix` is sufficient, so I wouldn't add this.

File: pkg/unpack/reflector.go [hunk 2]
```diff
@@ -222,6 +222,14 @@ func convertStruct(structPtr reflect.Value, in map[string]interface{}) error {
 		case reflect.Ptr:
 			derefType := emptyFieldVal.Type().Elem()
 			if derefType.Kind() == reflect.Struct {
+				subVal, ok := o.(reflect.Value)
+				if ok {
```
Comment (nwt) [inline pkg/unpack/reflector.go:226]:
```suggestion
				if subVal, ok := o.(reflect.Value); ok {
```

File: pkg/unpack/reflector.go [hunk 3]
```diff
@@ -222,6 +222,14 @@ func convertStruct(structPtr reflect.Value, in map[string]interface{}) error {
...
+					if subVal.Type().AssignableTo(emptyFieldVal.Type()) {
+						emptyFieldVal.Set(subVal)
+						continue
+					}
+					return fmt.Errorf("JSON field '%s': cannot assign value of type '%s' inside of struct type '%s'", fieldName, goName(subVal), goName(val))
```
Comment (nwt) [inline pkg/unpack/reflector.go:231]:
```suggestion
					return fmt.Errorf("JSON field %q: cannot assign value of type %q inside of struct type %q", fieldName, goName(subVal), goName(val))
```

**Review by nwt — APPROVED** [review]

File: compiler/ast/ast.go [hunk 4]
```diff
@@ -325,6 +344,32 @@ type Case struct {
...
+	Kind      string `json:"kind" unpack:""`
+	Keys      []Expr `json:"keys"`
+	Direction string `json:"direction"`
+}
+
+// In SqlJoin, Table is a type expression for a record that selects Z records with
+// a uniform schema and optionally places them in a sub-field of "."
+// according to Alias (which is a string expression).  Thus, exactly like
+// with SQL, you can refer to dotted references of columns using the
+// alias name.  Also, at some point the table expression can simply be a
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/ast/ast.go:362]: This comment is dated and this is all a work in progress so I will just delete it for this PR.

File: compiler/semantic/sql.go [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+		}
+		procs = append(procs, tableFilter)
+		if sql.From.Alias != nil {
+			// If there's an alias, we do a 'cut alias=.'
+			alias, err := convertSQLAlias(scope, sql.From.Alias)
+			if err != nil {
+				return nil, err
+			}
+			// If the FROM table has been aliased and all join clauses, if any,
+			// are also aliased, then we can lift any where claused that is dependent
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/semantic/sql.go:31]: same as above

File: compiler/semantic/sql.go [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+		if selector != nil {
+			procs = append(procs, selector)
+		}
+	}
+	if sql.OrderBy != nil {
+		direction := 1
+		if sql.OrderBy.Direction == "desc" {
+			direction = -1
+		}
+		procs = append(procs, sortByMulti(sql.OrderBy.Keys, direction))
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/semantic/sql.go:98]: See issue #2379

File: compiler/semantic/sql.go [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+		}
+		procs = append(procs, p)
+	}
+	if len(procs) == 0 {
+		procs = []ast.Proc{passProc}
+	}
+	return wrap(procs), nil
+}
+
+func isId(e ast.Expr) (string, bool) {
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/semantic/sql.go:113]: I use this pattern throughout the compiler and would prefer not to change it right now.  e.g., isCompareAny checks if the AST segment is a compare-any pattern, returns useful stuff, and an ok boolean.  We can change all of them later if this is confusing.

File: compiler/semantic/sql.go [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+
+func sqlField(scope *Scope, e ast.Expr) (field.Static, error) {
+	name, err := semField(scope, e)
+	if err != nil {
+		return nil, err
+	}
+	if f, ok := name.(*ast.Path); ok {
+		return f.Name, nil
+	}
+	return nil, errors.New("expression is not a field reference")
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/semantic/sql.go:510]: I'm not sure how to do this unless you mean call zfmt.

File: compiler/ztests/sql-join-groupby.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
+zql: |
+  SELECT p.name, union(f.name)
+  FROM fruit f
+  LEFT JOIN person p ON f.flavor = p.likes
+  GROUP BY p.name
+  | sort .
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/ztests/sql-join-groupby.yaml:6]: It's testing a mixture of Z and SQL.

File: field/field.go (hunk 1 above)
Comment (mccanne) [reply field/field.go:62]: It's not used so I will delete it but I found it useful at one point for code clarity at a call site.

File: compiler/semantic/sql.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nwt) [reply compiler/semantic/sql.go:113]: Works for me.

File: pkg/unpack/reflector.go (hunk 3 above)
Comment (mccanne) [reply pkg/unpack/reflector.go:231]: I will change this throughout.  This was written before  `%q` was in my repertoire :-)

File: compiler/semantic/sql.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (nwt) [reply compiler/semantic/sql.go:510]: Neither am I, so I suppose that's what I mean. Fine not to do this now, but for complex Z programs, we're eventually going to want a way to tie this sort of error back to the offending program text.

**Pull request merged**